File: src/firmware/c/main/modbus.c
#include "modbus.h"

uint16_t modbus_crc16(const uint8_t *data, size_t length)
{
    uint16_t crc = 0xFFFF;
    for (size_t i = 0; i < length; i++) {
        crc ^= (uint16_t)data[i];
        for (int bit = 0; bit < 8; bit++) {
            if (crc & 0x0001u) {
                crc = (uint16_t)((crc >> 1) ^ 0xA001u);
            } else {
                crc = (uint16_t)(crc >> 1);
            }
        }
    }
    return crc;
}

void modbus_build_read_holding(uint8_t slave_id,
                               uint16_t start_address,
                               uint16_t register_count,
                               uint8_t out[8])
{
    out[0] = slave_id;
    out[1] = MODBUS_FUNC_READ_HOLDING_REGISTERS;
    out[2] = (uint8_t)((start_address  >> 8) & 0xFFu);
    out[3] = (uint8_t)( start_address        & 0xFFu);
    out[4] = (uint8_t)((register_count >> 8) & 0xFFu);
    out[5] = (uint8_t)( register_count       & 0xFFu);
    uint16_t crc = modbus_crc16(out, 6);
    out[6] = (uint8_t)( crc        & 0xFFu);  /* CRC low byte first on the wire */
    out[7] = (uint8_t)((crc  >> 8) & 0xFFu);
}

modbus_status_t modbus_parse_read_holding(const uint8_t *resp,
                                          size_t resp_len,
                                          uint8_t expected_slave,
                                          const uint8_t **data_out,
                                          size_t *data_len_out)
{
    if (resp_len < 5) {
        return MODBUS_ERR_SHORT;
    }

    uint8_t slave    = resp[0];
    uint8_t func     = resp[1];

    /* Exception response: high bit of function code set; one byte exception code follows. */
    if (func & 0x80u) {
        return MODBUS_ERR_EXCEPTION;
    }
    if (func != MODBUS_FUNC_READ_HOLDING_REGISTERS) {
        return MODBUS_ERR_UNEXPECTED;
    }
    if (slave != expected_slave) {
        return MODBUS_ERR_UNEXPECTED;
    }

    uint8_t byte_count    = resp[2];
    size_t  expected_total = (size_t)3 + (size_t)byte_count + (size_t)2;
    if (resp_len < expected_total) {
        return MODBUS_ERR_SHORT;
    }

    /* Validate trailing CRC. */
    uint8_t  crc_lo  = resp[expected_total - 2];
    uint8_t  crc_hi  = resp[expected_total - 1];
    uint16_t reported = (uint16_t)((uint16_t)crc_hi << 8) | (uint16_t)crc_lo;
    uint16_t computed = modbus_crc16(resp, expected_total - 2);
    if (reported != computed) {
        return MODBUS_ERR_CRC;
    }

    *data_out     = &resp[3];
    *data_len_out = byte_count;
    return MODBUS_OK;
}

File: src/firmware/ts/modbus.ts
/**
 * Modbus RTU framing helpers — pure logic, no I/O.
 *
 * Function code 0x03 (read holding registers) is the only one we use.
 * Frame layout:
 *   [slaveId][0x03][addrHi][addrLo][countHi][countLo][crcLo][crcHi]
 * Response layout:
 *   [slaveId][0x03][byteCount][data...][crcLo][crcHi]
 */

export const FUNC_READ_HOLDING_REGISTERS = 0x03;

/** Modbus CRC-16 (polynomial 0xA001, init 0xFFFF). Returns a 16-bit value. */
export function crc16(bytes: Uint8Array, length = bytes.length): number {
  let crc = 0xFFFF;
  for (let i = 0; i < length; i++) {
    crc ^= bytes[i];
    for (let bit = 0; bit < 8; bit++) {
      if (crc & 0x0001) {
        crc = (crc >>> 1) ^ 0xA001;
      } else {
        crc >>>= 1;
      }
    }
  }
  return crc & 0xFFFF;
}

/** Build a "read holding registers" query frame with CRC appended. */
export function buildReadHoldingRegisters(
  slaveId:        number,
  startAddress:   number,
  registerCount:  number,
): Uint8Array {
  const frame = new Uint8Array(8);
  frame[0] = slaveId & 0xFF;
  frame[1] = FUNC_READ_HOLDING_REGISTERS;
  frame[2] = (startAddress  >> 8) & 0xFF;
  frame[3] =  startAddress        & 0xFF;
  frame[4] = (registerCount >> 8) & 0xFF;
  frame[5] =  registerCount       & 0xFF;
  const crc = crc16(frame, 6);
  frame[6] =  crc        & 0xFF;   // CRC low byte first
  frame[7] = (crc  >> 8) & 0xFF;
  return frame;
}

export interface ParsedResponse {
  slaveId:    number;
  functionId: number;
  /** Raw register payload (2 bytes per register, big-endian within each). */
  data:       Uint8Array;
}

export class ModbusError extends Error {
  constructor(message: string, public readonly code: 'SHORT' | 'CRC' | 'EXCEPTION' | 'UNEXPECTED') {
    super(message);
  }
}

/**
 * Parse a response frame. Throws ModbusError on framing/CRC errors.
 * Caller is responsible for further interpreting `data`.
 */
export function parseReadHoldingRegistersResponse(
  resp:           Uint8Array,
  expectedSlaveId: number,
): ParsedResponse {
  if (resp.length < 5) {
    throw new ModbusError(`Response too short (${resp.length} bytes)`, 'SHORT');
  }

  const slaveId    = resp[0];
  const functionId = resp[1];

  // Exception response: function code has high bit set; one byte exception code follows.
  if (functionId & 0x80) {
    throw new ModbusError(`Modbus exception code ${resp[2]}`, 'EXCEPTION');
  }

  if (functionId !== FUNC_READ_HOLDING_REGISTERS) {
    throw new ModbusError(`Unexpected function code 0x${functionId.toString(16)}`, 'UNEXPECTED');
  }

  if (slaveId !== expectedSlaveId) {
    throw new ModbusError(`Slave id mismatch (got ${slaveId}, expected ${expectedSlaveId})`, 'UNEXPECTED');
  }

  const byteCount = resp[2];
  const expectedTotal = 3 + byteCount + 2;
  if (resp.length < expectedTotal) {
    throw new ModbusError(`Response truncated (got ${resp.length}, expected ${expectedTotal})`, 'SHORT');
  }

  // Validate CRC.
  const crcLo = resp[expectedTotal - 2];
  const crcHi = resp[expectedTotal - 1];
  const reported = (crcHi << 8) | crcLo;
  const computed = crc16(resp, expectedTotal - 2);
  if (reported !== computed) {
    throw new ModbusError(`CRC mismatch (reported 0x${reported.toString(16)}, computed 0x${computed.toString(16)})`, 'CRC');
  }

  return {
    slaveId,
    functionId,
    data: resp.slice(3, 3 + byteCount),
  };
}
